### PR TITLE
nsenter waits for parent signal before forking

### PR DESCRIPTION
**Pull request by dqminh**
This allows the parent to place the process into cgroup first before spawning the user process.

### CODE REVIEW

**Comment by dqminh** [comment]
cc @crosbymichael 

There's also an integration test here https://github.com/docker/docker/pull/10122

File: namespaces/nsenter/nsenter.c [hunk 1]
```diff
@@ -151,6 +151,15 @@ void nsenter()
 			exit(1);
 		}
 	}
+	// read continue signal from fd 3. the signal is 1
+	// this is so that the parent can put the proces inside correct cgroups
+	// before it has a chance to fork
+	unsigned char s;
+	if (read(3, &s, 1) != 1 || s != '1') {
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: Should we check for EINTR on read here and retry?

Comment (dqminh) [inline namespaces/nsenter/nsenter.c]: I think that makes sense. EINTR is the only error that we want to retry right ?

File: namespaces/nsenter/nsenter.c [hunk 2]
```diff
@@ -151,6 +151,19 @@ void nsenter()
 			exit(1);
 		}
 	}
+	// read continue signal from fd 3. the signal is 1
+	// this is so that the parent can put the proces inside correct cgroups
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: typo: proces --> process.
Also, can we use something else besides the word "signal".
It is overloaded :)

Comment (dqminh) [inline namespaces/nsenter/nsenter.c]: haha yes !. How about `blocking until the parent placed the process inside correct cgroups.`. I think that's sufficient.

Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: Yes, that sounds better.

File: namespaces/nsenter/nsenter.c [hunk 3]
```diff
@@ -151,6 +151,17 @@ void nsenter()
...
 	}
+	// blocking until the parent placed the process inside correct cgroups.
+	unsigned char s;
+	ssize_t ret;
+	do {
+		ret = read(3, &s, 1);
+	} while (ret == -1 && errno == EINTR);
+	if (ret != 1 || s != '1') {
+		fprintf(stderr,
+			"nsenter: Failed to get signal from parent: %m\n");
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: Replace "signal" by "synchronization data" here?

Comment (dqminh) [inline namespaces/nsenter/nsenter.c]: fixed!

**Comment by mrunalp** [comment]
Thanks for making the changes. LGTM.

File: namespaces/nsenter/nsenter.c [hunk 4]
```diff
@@ -151,6 +151,17 @@ void nsenter()
...
 		}
 	}
+	// blocking until the parent placed the process inside correct cgroups.
+	unsigned char s;
+	ssize_t ret;
+	do {
+		ret = read(3, &s, 1);
+	} while (ret == -1 && errno == EINTR);
+	if (ret != 1 || s != '1') {
+		fprintf(stderr,
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: You should probably rebase and use the pr_perror macro from this commit https://github.com/docker/libcontainer/commit/35ae1c48710ff5a4db20645bc98c719cfb695b9a

File: namespaces/nsenter/nsenter.c [hunk 5]
```diff
@@ -151,6 +151,17 @@ void nsenter()
 			exit(1);
 		}
 	}
+	// blocking until the parent placed the process inside correct cgroups.
+	unsigned char s;
+	ssize_t ret;
+	do {
+		ret = read(3, &s, 1);
+	} while (ret == -1 && errno == EINTR);
```
Comment (avagin) [inline namespaces/nsenter/nsenter.c]: read() can return EINTR only if a process sets a signal handler without SA_RESTART. I think we can remove this loop from here. Thanks.

Comment (dqminh) [inline namespaces/nsenter/nsenter.c]: @avagin Thanks for the clarification! I agree that we can remove this loop here.

Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: Thanks for the pointer @avagin . I had a lingering suspicion that it might not be required in this case.

**Comment by crosbymichael** [comment]
@dqminh AWESOME.  Thanks for working on this!

**Comment by crosbymichael** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**